src/ui/public/vislib/lib/axis.js

```diff
@@ -0,0 +1,228 @@
+      opacity: 1,
+      tickColor: '#ddd',
+      tickWidth: '1px',
+      tickLength: '6px'
+    }
+  };
+
+  const categoryDefaults = {
+    type: 'category',
+    position: 'bottom',
+    labels: {
+      rotate: 0,
+      rotateAnchor: 'end',
+      filter: true
+    }
+  };
+  /**
+   * Appends y axis to the visualization
+   *
+   * @class Axis
+   * @constructor
+   * @param args {{el: (HTMLElement), yMax: (Number), _attr: (Object|*)}}
+   */
+  class Axis extends ErrorHandler {
+    constructor(args) {
+      super();
+      if (args.type === 'category') {
+        _.extend(this, defaults, categoryDefaults, args);
+      } else {
+        _.extend(this, defaults, args);
+      }
+
+      this._attr = args.vis._attr;
+      this.elSelector = this.elSelector.replace('{pos}', this.position);
+      this.scale = new AxisScale(this, {scale: this.scale});
+      this.axisTitle = new AxisTitle(this, this.axisTitle);
+      this.axisLabels = new AxisLabels(this, this.labels);
+    }
+
+    render() {
+      d3.select(this.vis.el).selectAll(this.elSelector).call(this.draw());
+    }
+
+    isHorizontal() {
+      return (this.position === 'top' || this.position === 'bottom');
+    }
+
+    getAxis(length) {
+      const scale = this.scale.getScale(length);
+
+      return d3.svg.axis()
+      .scale(scale)
+      .tickFormat(this.tickFormat(this.domain))
+      .ticks(this.tickScale(length))
+      .orient(this.position);
+    }
+
+    getScale() {
+      return this.scale.scale;
+    }
+
+    addInterval(interval) {
+      return this.scale.addInterval(interval);
+    }
+
+    substractInterval(interval) {
+      return this.scale.substractInterval(interval);
+    }
+
+    tickScale(length) {
+      const yTickScale = d3.scale.linear()
+      .clamp(true)
+      .domain([20, 40, 1000])
+      .range([0, 3, 11]);
+
+      return Math.ceil(yTickScale(length));
+    }
+
+    tickFormat() {
+      if (this.axisFormatter) return this.axisFormatter;
+      if (this.isPercentage()) return d3.format('%');
+      return d3.format('n');
+    }
+
+    getLength(el, n) {
+      if (this.isHorizontal()) {
+        return $(el).parent().width() / n - this._attr.margin.left - this._attr.margin.right - 50;
+      }
+      return $(el).parent().height() / n - this._attr.margin.top - this._attr.margin.bottom;
+    }
+
+    updateXaxisHeight() {
+      const self = this;
+      const selection = d3.select(this.vis.el).selectAll('.vis-wrapper');
+
+
+      selection.each(function () {
+        const visEl = d3.select(this);
+
+        if (visEl.select('.inner-spacer-block').node() === null) {
+          visEl.selectAll('.y-axis-spacer-block')
+          .append('div')
+          .attr('class', 'inner-spacer-block');
+        }
+
+        const height = visEl.select(`.axis-wrapper-${self.position}`).style('height');
+        visEl.selectAll(`.y-axis-spacer-block-${self.position} .inner-spacer-block`).style('height', height);
+      });
+    }
+
+    adjustSize() {
```

Comment:
Is the naming of functions like this and `draw()` normal to you? I would expect this and `draw()` to be called something like `getDrawFunction()` and `getAdjustSizeFunction()`.

That said, I also don't understand why these functions don't just take the selection and do the job.